crates/cheatcodes/src/evm.rs

```diff
@@ -471,17 +470,67 @@ impl Cheatcode for readCallersCall {
     }
 }
 
-impl Cheatcode for snapshotCall {
+impl Cheatcode for lastCallGasCall {
+    fn apply(&self, state: &mut Cheatcodes) -> Result {
+        let Self {} = self;
+        let Some(last_call_gas) = &state.last_call_gas else {
+            bail!("no external call was made yet");
+        };
+        Ok(last_call_gas.abi_encode())
+    }
+}
+
+impl Cheatcode for startSnapshotGasCall {
+    fn apply_stateful<DB: DatabaseExt>(&self, ccx: &mut CheatsCtxt<DB>) -> Result {
+        let Self { name } = self;
+        let record = GasRecord { name: name.clone(), gas: ccx.state.last_call_gas.clone() };
+        ccx.state.recorded_gas.push(record);
+        Ok(Default::default())
+    }
+}
+
+impl Cheatcode for stopSnapshotGasCall {
+    fn apply_stateful<DB: DatabaseExt>(&self, ccx: &mut CheatsCtxt<DB>) -> Result {
+        let Self { name } = self;
+        let record = ccx
+            .state
+            .recorded_gas
+            .iter_mut()
+            .find(|record| record.name == *name)
+            .ok_or_else(|| fmt_err!("gas snapshot not found: {name}"))?;
+        let gas_end = ccx.state.last_call_gas.clone();
+
+        let gas_used = gas_end
+            .as_ref()
+            .zip(record.gas.as_ref())
+            .map(|(end, start)| end.gasTotalUsed - start.gasTotalUsed)
+            .unwrap_or_default();
+
+        // Create the snapshot if it doesn't exist
+        create_dir_all(ccx.state.config.paths.snapshots.clone())?;
+
+        // Write the snapshot to a file
+        let snapshot_path = ccx.state.config.paths.snapshots.join(name).join(".json");
```

Comment:
would be nice to have a check for multiple snapshots with the same name (common mistake causing confusing snapshot thrashing based on test runner ordering)

Reply:
recognize this is probably harder but would love if name could be optional and autogenerated as `testContract.testFunction`

Reply:
Supportive, I think it should be possible

Reply:
I was able to add support for generating group names by contract name but haven't figured out a way to support derived function names yet. This can be added as a follow-up PR in a non-breaking way.